lib/MySQL_Thread.cpp

```diff
@@ -1464,3 +1872,30 @@ void MySQL_Threads_Handler::signal_all_threads() {
 	}
 }
 
+bool MySQL_Threads_Handler::kill_session(uint32_t _thread_session_id) {
+	bool ret=false;
+	unsigned int i;
+	signal_all_threads(1);
+	for (i=0;i<num_threads;i++) {
```

Comment:
It seems to me that 'thread_mutex' is being used all the time in order to lock all threads. Because it's always more dangerous to have more locks than a single lock, why not have a single global lock for starters?

Effectively, this set of mutexes do just that. But when you have more than one mutex, you always have to make sure to lock and release them in the correct order.

BTW, I think the order of release is wrong and should be the reverse order.

All in all, when it comes to locking, we don't want deadlocks in the proxy :)